api-authentication/server/index.js

```diff
@@ -0,0 +1,111 @@
+require('dotenv/config');
+const pg = require('pg');
+const argon2 = require('argon2'); // eslint-disable-line
+const express = require('express');
+const jwt = require('jsonwebtoken'); // eslint-disable-line
+const ClientError = require('./client-error');
+const errorMiddleware = require('./error-middleware');
+
+const db = new pg.Pool({ // eslint-disable-line
+  connectionString: process.env.DATABASE_URL,
+  ssl: {
+    rejectUnauthorized: false
+  }
+});
+
+const app = express();
+
+const jsonMiddleware = express.json();
+
+app.use(jsonMiddleware);
+
+app.post('/api/auth/sign-up', (req, res, next) => {
+  const { username, password } = req.body;
+  if (!username || !password) {
+    throw new ClientError(400, 'username and password are required fields');
+  }
+  argon2
+    .hash(password)
+    .then(hashedPassword => {
+      const sql = `
+        insert into "users" ("username", "hashedPassword")
+        values ($1, $2)
+        returning "userId", "username", "createdAt"
+      `;
+      const params = [username, hashedPassword];
+      return db.query(sql, params);
+    })
+    .then(result => {
+      const [user] = result.rows;
+      res.status(201).json(user);
+    })
+    .catch(err => next(err));
+});
+
+app.post('/api/auth/sign-in', (req, res, next) => {
+  const { username, password } = req.body;
+  if (!username || !password) {
+    throw new ClientError(401, 'invalid login');
+  }
+  /* your code starts here */
+  const sql = `
+          SELECT "userId",
+              "hashedPassword"
+        from "users"
+        where "username" = $1;
+      `;
+  const params = [username];
+  db.query(sql, params)
+    .then(result => {
+      const [userlogin] = result.rows;
+      if (!userlogin) {
+        throw new ClientError(401, 'invalid login');
+      } else {
+        argon2
+          .verify(userlogin.hashedPassword, password)
+          .then(isMatching => {
+            if (!isMatching) {
+              throw new ClientError(401, 'invalid login');
+            } else {
+              const payload = {
+                userId: userlogin.userId,
+                username
```

Comment:
It is best and safest to use the data from the database when available instead of the user inputted data. So here it would be better to do `username: userLogin.username`, and update the query so you have the `username` in your database results.